src/llmq/utils.cpp

```diff
@@ -108,6 +108,98 @@ static uint256 GetHashModifier(const Consensus::LLMQParams& llmqParams, gsl::not
     return ::SerializeHash(std::make_pair(llmqParams.type, pCycleQuorumBaseBlockIndex->GetBlockHash()));
 }
 
+static arith_uint256 calculateQuorumScore(const CDeterministicMNCPtr& dmn, const uint256& modifier)
+{
+    // calculate sha256(sha256(proTxHash, confirmedHash), modifier) per MN
+    // Please note that this is not a double-sha256 but a single-sha256
+    // The first part is already precalculated (confirmedHashWithProRegTxHash)
+    // TODO When https://github.com/bitcoin/bitcoin/pull/13191 gets backported, implement something that is similar but for single-sha256
+    uint256 h;
+    CSHA256 sha256;
+    sha256.Write(dmn->pdmnState->confirmedHashWithProRegTxHash.begin(),
+                 dmn->pdmnState->confirmedHashWithProRegTxHash.size());
+    sha256.Write(modifier.begin(), modifier.size());
+    sha256.Finalize(h.begin());
+    return UintToArith256(h);
+}
+
+static std::vector<std::pair<arith_uint256, CDeterministicMNCPtr>> CalculateScoresForQuorum(
+    std::vector<CDeterministicMNCPtr>&& dmns, const uint256& modifier, const bool onlyEvoNodes)
+{
+    std::vector<std::pair<arith_uint256, CDeterministicMNCPtr>> scores;
+    scores.reserve(dmns.size());
+
+    for (auto& dmn : dmns) {
+        if (dmn->pdmnState->IsBanned()) continue;
+        if (dmn->pdmnState->confirmedHash.IsNull()) {
+            // we only take confirmed MNs into account to avoid hash grinding on the ProRegTxHash to sneak MNs into a
+            // future quorums
+            continue;
+        }
+        if (onlyEvoNodes && dmn->nType != MnType::Evo) {
+            continue;
+        }
+        scores.emplace_back(calculateQuorumScore(dmn, modifier), std::move(dmn));
+    };
+    return scores;
+}
+
+static std::vector<std::pair<arith_uint256, CDeterministicMNCPtr>> CalculateScoresForQuorum(
+    const CDeterministicMNList& mn_list, const uint256& modifier, const bool onlyEvoNodes)
+{
+    std::vector<std::pair<arith_uint256, CDeterministicMNCPtr>> scores;
+    scores.reserve(mn_list.GetAllMNsCount());
+
+    mn_list.ForEachMNShared(true, [&](const CDeterministicMNCPtr& dmn) {
+        if (dmn->pdmnState->confirmedHash.IsNull()) {
+            // we only take confirmed MNs into account to avoid hash grinding on the ProRegTxHash to sneak MNs into a
+            // future quorums
+            return;
+        }
+        if (onlyEvoNodes && dmn->nType != MnType::Evo) {
+            return;
+        }
+
+        scores.emplace_back(calculateQuorumScore(dmn, modifier), dmn);
+    });
+    return scores;
+}
+
+
+/**
+ * Calculate a quorum based on the modifier. The resulting list is deterministically sorted by score
+ */
+template <typename List>
+static std::vector<CDeterministicMNCPtr> CalculateQuorum(List&& mn_list, const uint256& modifier, size_t maxSize = 0,
```

Comment:
It seems to me better to make this accept a Span<CDeterministicMN> so that we have non-templated types